src/miner_blockhistory.erl

```diff
@@ -0,0 +1,97 @@
+-module(miner_blockhistory).
+
+-behavior(gen_server).
+
+-record(state, {
+    url,
+    historical_block=0,
+    historical_time=0,
+    last_retrieved_time=0
+}).
+
+-define(HISTORICAL_TIME_FETCH_INTERVAL, 3600).
+
+-export([init/1, handle_info/2, handle_cast/2, handle_call/3]).
+
+-export([start_link/0, get_history/0]).
+
+start_link() ->
+    BaseURL = application:get_env(miner, api_base_url, "https://api.helium.io/v1"),
+    URL = string:concat(BaseURL, "/blocks/"),
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [URL], []).
+
+-spec get_history() -> {non_neg_integer(), non_neg_integer()} | undefined.
+get_history() ->
+    gen_server:call(?MODULE, get_history).
+
+%% gen_server callbacks
+
+init([URL]) ->
+    {ok, schedule_check(#state{url=URL}), 15}.
+
+handle_info(check, #state{url=URL}=State) ->
+    case blockchain_worker:blockchain() of
+        undefined ->
+            lager:info("No chain, retry soon"),
+            {ok, schedule_check(State, 15)};
+        Chain ->
+            Ledger = blockchain:ledger(Chain),
+            case blockchain_ledger_v1:current_height(Ledger) of
+                {ok, CurrentHeight} ->
+                    StabilizationPeriod = application:get_env(miner, stabilization_period, 50000),
+                    HistoryHeight = max(1, CurrentHeight - StabilizationPeriod),
+                    HeightURL = string:concat(URL, integer_to_list(HistoryHeight)),
+                    case httpc:request(get, {HeightURL, [{"user-agent", "ValidatorHistoricalHeight/0.0.1"}] }, [], [{body_format, binary}]) of
+                        {ok, {{_HttpVersion, 200, "OK"}, _Headers, Body}} ->
+                            try jsx:decode(Body, [{return_maps, true}]) of
+                                Json ->
+                                    case maps:get(<<"data">>, Json, undefined) of
```

Comment:
You could probably do `maps:get(<<"time">>, maps:get(<<"data">>, Json, #{}), undefined)` and save yourself a case statement here